File: regex.py
def QuestionMark(pattern, text):
    # Parçalanmış desenler için boş bir liste başlatır.
    splitPattern = []
    # Eşleşmeleri saklamak için boş bir liste başlatır.
    matches = []
    # Desenin uzunluğunu alır.
    lenPattern = len(pattern)
    
    # Deseni incelemek için döngü oluşturur.
    for i in range(lenPattern - 1):
        # Eğer 'pattern' içinde soru işareti ('?') varsa
        if pattern[i] == "?":
            # Soru işaretinden önceki kısmı 'splitPattern'e ekler.
            splitPattern.append(pattern[:i - 1])
            # Soru işaretinden sonraki kısmı 'splitPattern'e ekler.
            splitPattern.append(pattern[i + 1:])
            # Bölümleri birleştirir ve 'matches' listesine ekler.
            adder = "".join(splitPattern)
            matches.append(adder)

            # Listeyi temizleyerek baştan başlar.
            splitPattern = []
            # Soru işaretinden hemen önceki kısmı 'splitPattern'e ekler.
            splitPattern.append(pattern[:i])
            # Soru işaretinden hemen sonraki kısmı 'splitPattern'e ekler.
            splitPattern.append(pattern[i + 1:])
            # Bölümleri birleştirir ve 'matches' listesine ekler.
            adder = "".join(splitPattern)
            matches.append(adder)

    # 'matches' listesindeki parçaları kontrol etmek için döngü.
    i = 0
    while i < len(matches):
        # Metin içinde mevcut parçayı kontrol eder ve bulursa yazdırır.
        if matches[i] in text:
            print(matches[i])
        # Bir sonraki parçaya geçer.
        i += 1

def OR(pattern, text):
    # Deseni '|' ile bölmek için boş bir liste tanımlıyoruz.
    splitPattern = []
    # Eşleşmeleri saklamak için boş bir liste tanımlıyoruz.
    matches = []
    # Desenin uzunluğunu hesaplıyoruz.
    lenPattern = len(pattern)

    # Desen içinde '|' olup olmadığını kontrol ediyoruz ve varsa bölüyoruz.
    for i in range(lenPattern - 1):
        if pattern[i] == "|":
            splitPattern = pattern.split("|")

    # Bölünmüş desenin her bir parçasını 'matches' listesine ekliyoruz.
    for i in range(len(splitPattern)):
        matches.append(splitPattern[i])

    # 'matches' listesini döngüyle kontrol ederek eşleşmeleri buluyoruz.
    i = 0
    while i < len(matches):
        # Metin içinde 'matches' listesindeki parçanın bulunup bulunmadığını kontrol ediyoruz.
        if matches[i] in text:
            # Eşleşme bulunduğunda, bu parçayı yazdırıyoruz.
            print(matches[i])
        # Bir sonraki elemanı kontrol etmek için ilerliyoruz.
        i += 1

def PlusSign(pattern, text):
    # '+' işaretini baz alarak deseni parçalara ayırırız.
    splitPattern = pattern.split("+")
    start_sequence = splitPattern[0]  # '+' işaretinden önceki kısım
    endSequence = splitPattern[1]  # '+' işaretinden sonraki kısım

    matches = []
    # Metinde eşleşmeleri bulmak için indeks kullanırız.
    i = 0
    while i < len(text):
        # Metnin başlangıcı artı işaretinden önceki kısımla eşleşiyor mu?
        if text[i:i+len(start_sequence)] == start_sequence:
            # '+' işareti anlamındaki ardışık karakterleri kontrol ederiz.
            j = i + len(start_sequence)
            while j < len(text) and text[j] == start_sequence[-1]:
                j += 1
            # '+' işaretinden sonraki kısmı kontrol ederiz.
            if text[j:j+len(endSequence)] == endSequence:
                # Tam bir eşleşme bulunduğunda, bu kısmı yazdırırız.
                matches.append(text[i:j + len(endSequence)])
                # Bir sonraki arama için indeksi ayarlayarak devam ederiz.
                i = j + len(endSequence) - 1
            else:
                i += 1
        else:
            i += 1
    for i in matches:
        print(i)

def StarSign(pattern, text):
    # '*' işaretini baz alarak deseni parçalara ayırırız.
    splitPattern = pattern.split("*")
    startSequence = splitPattern[0]  # '*' işaretinden önceki kısım
    endSequence = splitPattern[1]  # '*' işaretinden sonraki kısım
    beforeStar = startSequence[-1] # '*' işaretinden önceki index
    
    matches = []
    firstMatch = (startSequence + endSequence).split(beforeStar)
    firstMatch = "".join(firstMatch)
    matches.append(firstMatch) #'*' işareti olmayan durumun array'e eklenmesi
    
    i = 0
    while i < len(text):
        # Metnin başlangıcı start_sequence ile eşleşiyor mu?
        if text[i:i + len(startSequence)] == startSequence:
            # Başlangıç sekansı ile eşleşme bulduğumuzda, yıldız işaretinden sonraki karakterleri kontrol ederiz
            j = i + len(startSequence)
            while j < len(text) and text[j] == startSequence[-1]:
                j += 1  # Yıldızdan sonra gelen karakterleri kontrol ediyoruz
            # Bitiş sekansı ile eşleşiyor mu?
            if text[j:j + len(endSequence)] == endSequence:
                # Eşleşmeyi bulduğumuzda, matches dizisine ekleriz
                matches.append(text[i:j + len(endSequence)])
                # Bir sonraki kontrol için indeksi güncelleriz
                i = j + len(endSequence)
            else:
                i += 1  # Eşleşme yoksa, bir adım ilerleriz
        else:
            i += 1  # Başlangıç sekansı ile eşleşmezse, bir adım ilerleriz
    
    for i in matches:
        print(i)

def CurlyBracket(pattern,text):
    matches = []

    # Text'i boşluklara göre ayırırız.
    textElements = text.split()

    # Start ve end sekanslarını ayırırız
    if '{' in pattern and '}' in pattern:
        # Deseni {} arasındaki kısmı ayırmak için kullanılır
        startEndSequence = pattern.split('}')
        startSequence = startEndSequence[0].split('{')[0]
        repeatValues = startEndSequence[0].split('{')[1]

        # Virgülle ayırdıktan sonra değerleri alırız
        rangeValues = repeatValues.split(',')

        # Aralık belirtilmişse
        if len(rangeValues) == 2:
            start_value = int(rangeValues[0])
            # Eğer aralığın sonu boşsa, varsayılan bir maksimum tekrar belirleriz
            if rangeValues[1] == '':
                max_repeat = 20
                for i in range(start_value, max_repeat + 1):
                    new_pattern = startSequence * i
                    if new_pattern in textElements:
                        matches.append(new_pattern)
            else:
                end_value = int(rangeValues[1])
                # Aralığın başlangıcından bitişine kadar tekrarlama yaparız
                for i in range(start_value, end_value + 1):
                    new_pattern = startSequence * i
                    if new_pattern in textElements:
                        matches.append(new_pattern)

        elif len(rangeValues) == 1:
            # Belirtilen tek bir tekrar varsa
            new_pattern = startSequence * int(rangeValues[0])
            if new_pattern in textElements:
                matches.append(new_pattern)
    
    print(matches)

def SquareBracket(pattern, text):
    # Desende kare parantez içindeki deseni ayıklayın
    insideBrackets = pattern.split("[")[1].split("]")[0]
    # Kare parantezlerin dışındaki başlangıç ve bitiş desenlerini alın
    startPattern = pattern.split("[")[0]
    endPattern = pattern.split("]")[1]

    # Metni kelimelere ayırın
    splittedText = text.split()
    # Eşleşmeleri kaydetmek için liste oluşturun
    matches = []

    # Metin içindeki kelimeler arasında kontrol yapın
    for word in splittedText:
        # Başlangıç ve bitiş desenlerine uyan kelimeleri kontrol edin
        if word.startswith(startPattern) and word.endswith(endPattern):
            # Kare parantez içindeki desene uygun olup olmadığını kontrol edin
            if insideBrackets in word:
                matches.append(word)

    # Eşleşmeleri yazdırın
    print(matches)

def Bracket(pattern, text):
    # Yuvarlak parantez içindeki ifadeleri toplamak için liste
    matches = []
    # Parantezler arasında olup olmadığını kontrol etmek için bayrak
    inParentheses = False
    # Parantez içindeki karakterleri toplamak için değişken
    temp = ""

    splittedText = text.split() 

    # Deseni karakter bazında döngüyle tarayın
    for char in pattern:
        if char == '(':  # Parantez başlarsa, bayrağı açın ve temizleyin
            inParentheses = True
            temp = ""  # Temizleyin
        elif char == ')':  # Parantez kapanırsa, bayrağı kapatın ve sonucu ekleyin
            if inParentheses:
                matches.append(temp)
                inParentheses = False
        elif inParentheses:  # Parantez içindeyken karakterleri toplayın
            temp += char

    # Elde edilen ifadeleri metin içinde kontrol edin ve eşleşmeleri yazdırın
    for result in matches:
        if result in splittedText:
            print(matches)

def PowerSign(pattern, text):
    # 'text' kelimelerini ayırmak için boşlukla bölüyoruz
    splittedText = text.split()
    # 'pattern' içinde '^' karakterinden sonraki kısmı alıyoruz
    start_pattern = pattern.split("^")[1]
    # Eşleşen kelimeleri tutmak için liste oluşturuyoruz
    matches = []

    # Tüm kelimeleri kontrol ediyoruz
    for word in splittedText:
        # Eğer kelime başlangıç paternine uyuyorsa, listeye ekliyoruz
        if word.startswith(start_pattern):
            matches.append(word)
    
    # Eşleşen kelimeleri yazdırıyoruz
    print(matches)

def DollarSign(pattern, text):
    # Metindeki kelimeleri ayırmak için boşlukla bölüyoruz
    splittedText = text.split()
    # 'pattern' içindeki '$' işaretinden önceki kısmı alıyoruz
    end_pattern = pattern.split("$")[0]
    # Eşleşen kelimeleri tutmak için liste oluşturuyoruz
    matches = []

    # Metindeki her kelimeyi kontrol ediyoruz
    for word in splittedText:
        # Eğer kelime bitiş paternine uyuyorsa, listeye ekliyoruz
        if word.endswith(end_pattern):
            matches.append(word)
    
    # Eşleşen kelimeleri yazdırıyoruz
    print(matches)

def PointSign(pattern, text):
    # Metindeki kelimeleri boşluklarla ayırıyoruz
    words = text.split()

    # 'pattern' içindeki '.' işaretine göre bölüyoruz
    split_pattern = pattern.split(".")
    start_pattern = split_pattern[0]  # '.' işaretinden önceki kısım
    end_pattern = split_pattern[1]  # '.' işaretinden sonraki kısım

    matches = []

    # Tüm kelimeleri kontrol ediyoruz
    for word in words:
        # Başlangıç ve bitiş desenlerine uyan ve uzunluğu aynı olan kelimeleri seçiyoruz
        if word.startswith(start_pattern) and word.endswith(end_pattern) and len(word) == len(pattern):
            matches.append(word)

    # Eşleşen kelimeleri yazdırıyoruz
    print(matches)

def parser(pattern, text):
    # Desendeki her bir karakteri kontrol ederiz
    for i in pattern:
        # Eğer karakter metakarakterler listesinde ise uygun fonksiyonu çağırırız
        if i in metacharacters:
            if i == "?":
                QuestionMark(pattern, text)
            elif i == "|":
                OR(pattern, text)
            elif i == "*":
                StarSign(pattern, text)
            elif i == "+":
                PlusSign(pattern, text)
            elif i == "{":
                CurlyBracket(pattern, text)
            elif i == "[":
                SquareBracket(pattern, text)
            elif i == "(":
                Bracket(pattern, text)
            elif i == "^":
                PowerSign(pattern, text)
            elif i == "$":
                DollarSign(pattern, text)
            elif i == ".":
                PointSign(pattern, text)

pattern =input("Lütfen Pattern giriniz: ")
text =input("Lütfen Text giriniz: ")

# Regex metacharacters listesi
metacharacters = ["?", ".", "*", "+", "^", "$", "|", "(", ")", "[", "]", "{", "}", "\\"]

parser(pattern,text)

File: regex.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <algorithm>

void QuestionMark(const std::string &pattern, const std::string &text)
{
    // Deseni bölmek için boş bir vektör oluştur
    std::vector<std::string> splitPattern;
    // Eşleşmeleri saklamak için bir vektör oluştur
    std::vector<std::string> matches;
    // Desenin uzunluğunu al
    size_t lenPattern = pattern.size();

    // Deseni incelemek için döngü oluştur
    for (size_t i = 0; i < lenPattern - 1; ++i)
    {
        // Eğer desen içinde soru işareti varsa
        if (pattern[i] == '?')
        {
            // Soru işaretinin öncesini ekle
            splitPattern.push_back(pattern.substr(0, i - 1));
            // Soru işaretinin sonrasını ekle
            splitPattern.push_back(pattern.substr(i + 1));

            // Bölümleri birleştir ve eşleşmelere ekle
            std::string adder = "";
            for (const auto &part : splitPattern)
            {
                adder += part;
            }
            matches.push_back(adder);

            // SplitPattern'i temizle ve yeniden başla
            splitPattern.clear();

            // Soru işaretinin hemen öncesini ve sonrasını ekle
            splitPattern.push_back(pattern.substr(0, i));
            splitPattern.push_back(pattern.substr(i + 1));

            // Bölümleri birleştir ve eşleşmelere ekle
            adder = "";
            for (const auto &part : splitPattern)
            {
                adder += part;
            }
            matches.push_back(adder);
        }
    }

    // Eşleşmeleri kontrol etmek için bir döngü
    for (size_t i = 0; i < matches.size(); ++i)
    {
        // Eğer metin içinde eşleşme varsa, yazdır
        if (text.find(matches[i]) != std::string::npos)
        {
            std::cout << matches[i] << std::endl;
        }
    }
}

void OR(const std::string &pattern, const std::string &text)
{
    // Deseni '|' sembolü ile bölecek bir vektör oluştur.
    std::vector<std::string> splitPattern;
    // Eşleşmeleri saklamak için bir vektör oluştur.
    std::vector<std::string> matches;
    // Desenin uzunluğunu al.
    size_t lenPattern = pattern.size();

    // Deseni '|' sembolü ile böler ve 'splitPattern' içine ekler.
    std::istringstream stream(pattern);
    std::string segment;
    while (std::getline(stream, segment, '|'))
    {
        splitPattern.push_back(segment);
    }

    // 'splitPattern' içindeki her bir segmenti 'matches' listesine ekler.
    for (const auto &part : splitPattern)
    {
        matches.push_back(part);
    }

    // Eşleşmeleri kontrol etmek için bir döngü.
    for (const auto &match : matches)
    {
        // Metin içinde eşleşmeyi kontrol eder.
        if (text.find(match) != std::string::npos)
        {
            // Eğer eşleşme varsa, konsola yazdırır.
            std::cout << match << std::endl;
        }
    }
}

void PlusSign(const std::string &pattern, const std::string &text)
{
    // '+' işaretine göre deseni parçalara ayır.
    size_t plusPos = pattern.find('+');
    std::string startSequence = pattern.substr(0, plusPos); // '+' işaretinden önceki kısım
    std::string endSequence = pattern.substr(plusPos + 1);  // '+' işaretinden sonraki kısım

    std::vector<std::string> matches;

    // Metinde eşleşmeleri bulmak için bir döngü.
    size_t i = 0;
    while (i < text.size())
    {
        // Metin içinde başlangıç kısmıyla eşleşme olup olmadığını kontrol et.
        if (text.substr(i, startSequence.size()) == startSequence)
        {
            // '+' işaretine göre ardışık karakterleri kontrol et.
            size_t j = i + startSequence.size();
            while (j < text.size() && text[j] == startSequence.back())
            {
                j++;
            }
            // '+' işaretinden sonraki kısımla eşleşme olup olmadığını kontrol et.
            if (text.substr(j, endSequence.size()) == endSequence)
            {
                // Tam bir eşleşme bulunduğunda, bu kısmı ekle.
                matches.push_back(text.substr(i, j - i + endSequence.size()));
                // İndeksi güncelleyerek devam et.
                i = j + endSequence.size() - 1;
            }
            else
            {
                i++;
            }
        }
        else
        {
            i++;
        }
    }

    // Eşleşmeleri konsolda yazdır.
    for (const auto &match : matches)
    {
        std::cout << match << std::endl;
    }
}

void StarSign(const std::string &pattern, const std::string &text)
{
    // '*' sembolünün pozisyonunu bul
    size_t starPos = pattern.find('*');
    if (starPos == std::string::npos)
    {
        std::cerr << "Error: Pattern does not contain '*'" << std::endl;
        return;
    }

    // Başlangıç ve bitiş sekanslarını ayır
    std::string startSequence = pattern.substr(0, starPos);
    std::string endSequence = (starPos + 1 < pattern.size()) ? pattern.substr(starPos + 1) : "";

    if (startSequence.empty())
    {
        std::cerr << "Error: Start sequence is empty." << std::endl;
        return;
    }

    char beforeStar = startSequence.back(); // '*' işaretinden önceki karakter
    std::vector<std::string> matches;

    // Yıldız olmadan ilk eşleşmeyi oluştur
    std::string firstMatch = startSequence + endSequence;
    firstMatch.erase(std::remove(firstMatch.begin(), firstMatch.end(), beforeStar), firstMatch.end());
    matches.push_back(firstMatch);

    // Metin içinde sekansları kontrol et
    size_t i = 0;
    while (i < text.size())
    {
        if (i + startSequence.size() > text.size())
        {
            break; // StartSequence dışına çıkılmasın
        }

        // Başlangıç sekansını kontrol et
        if (text.substr(i, startSequence.size()) == startSequence)
        {
            size_t j = i + startSequence.size();

            // Yıldızdan önceki karakteri atla
            while (j < text.size() && text[j] == beforeStar)
            {
                j++;
            }

            // Bitiş sekansını kontrol et
            if (j + endSequence.size() <= text.size() &&
                text.substr(j, endSequence.size()) == endSequence)
            {
                matches.push_back(text.substr(i, j - i + endSequence.size()));
                i = j + endSequence.size(); // İndeksi ileri taşımak
            }
            else
            {
                i++;
            }
        }
        else
        {
            i++;
        }
    }

    // Eşleşmeleri konsola yazdır
    for (const auto &match : matches)
    {
        std::cout << match << std::endl;
    }
}

void CurlyBracket(const std::string &pattern, const std::string &text)
{
    std::vector<std::string> matches;

    // Metni boşluklara göre ayır.
    std::istringstream textStream(text);
    std::vector<std::string> textElements;
    std::string word;
    while (textStream >> word)
    {
        textElements.push_back(word);
    }

    // Start ve end sekanslarını belirle
    size_t openBracePos = pattern.find('{');
    size_t closeBracePos = pattern.find('}');

    if (openBracePos != std::string::npos && closeBracePos != std::string::npos)
    {
        std::string startSequence = pattern.substr(0, openBracePos);
        std::string repeatValues = pattern.substr(openBracePos + 1, closeBracePos - openBracePos - 1);

        std::vector<std::string> rangeValues;
        std::istringstream repeatStream(repeatValues);
        std::string token;
        while (std::getline(repeatStream, token, ','))
        {
            rangeValues.push_back(token);
        }

        if (rangeValues.size() == 2)
        {
            int start_value = std::stoi(rangeValues[0]);
            int max_repeat = rangeValues[1].empty() ? 20 : std::stoi(rangeValues[1]);

            for (int i = start_value; i <= max_repeat; i++)
            {
                // String'i elle tekrarla
                std::string new_pattern = "";
                for (int j = 0; j < i; j++)
                {
                    new_pattern += startSequence; // String'i j kadar tekrarla
                }

                if (std::find(textElements.begin(), textElements.end(), new_pattern) != textElements.end())
                {
                    matches.push_back(new_pattern);
                }
            }
        }
        else if (rangeValues.size() == 1)
        {
            int repeat = std::stoi(rangeValues[0]);

            // String'i elle tekrarla
            std::string new_pattern = "";
            for (int j = 0; j < repeat; j++)
            {
                new_pattern += startSequence; // String'i j kadar tekrarla
            }

            if (std::find(textElements.begin(), textElements.end(), new_pattern) != textElements.end())
            {
                matches.push_back(new_pattern);
            }
        }
    }

    // Eşleşmeleri konsola yazdır.
    for (const auto &match : matches)
    {
        std::cout << match << std::endl;
    }
}

void SquareBracket(const std::string &pattern, const std::string &text)
{
    // Kare parantez içindeki deseni çıkar
    size_t openBracket = pattern.find('[');
    size_t closeBracket = pattern.find(']');

    std::string insideBrackets = pattern.substr(openBracket + 1, closeBracket - openBracket - 1);
    std::string startPattern = pattern.substr(0, openBracket);
    std::string endPattern = pattern.substr(closeBracket + 1);

    // Metni kelimelere ayırın
    std::istringstream textStream(text);
    std::vector<std::string> splittedText;
    std::string word;
    while (textStream >> word)
    {
        splittedText.push_back(word);
    }

    // Eşleşmeleri kaydetmek için bir liste oluşturun
    std::vector<std::string> matches;

    // Metin içindeki kelimeleri kontrol edin
    for (const auto &word : splittedText)
    {
        // Başlangıç ve bitiş desenlerine uygun olup olmadığını kontrol edin
        if (word.find(startPattern) == 0 && word.rfind(endPattern) == (word.size() - endPattern.size()))
        {
            // Kare parantez içindeki desene uygun olup olmadığını kontrol edin
            if (word.find(insideBrackets) != std::string::npos)
            {
                matches.push_back(word);
            }
        }
    }

    // Eşleşmeleri yazdırın
    for (const auto &match : matches)
    {
        std::cout << match << std::endl;
    }
}

void Bracket(const std::string &pattern, const std::string &text)
{
    // Yuvarlak parantez içindeki ifadeleri toplamak için liste
    std::vector<std::string> matches;
    // Parantezler arasında olup olmadığını kontrol etmek için bayrak
    bool inParentheses = false;
    // Parantez içindeki karakterleri toplamak için geçici bir değişken
    std::string temp;

    // Metni kelimelere ayırın
    std::istringstream textStream(text);
    std::vector<std::string> splittedText;
    std::string word;
    while (textStream >> word)
    {
        splittedText.push_back(word);
    }

    // Deseni karakter bazında döngüyle tarayın
    for (const char &charac : pattern)
    {
        if (charac == '(')
        { // Parantez başlarsa
            inParentheses = true;
            temp.clear(); // Geçici değişkeni temizleyin
        }
        else if (charac == ')')
        { // Parantez kapanırsa
            if (inParentheses)
            {
                matches.push_back(temp); // Sonucu listeye ekleyin
                inParentheses = false;   // Bayrağı kapatın
            }
        }
        else if (inParentheses)
        { // Parantez içindeyken karakterleri toplayın
            temp += charac;
        }
    }

    // Elde edilen ifadeleri metin içinde kontrol edin ve eşleşmeleri yazdırın
    for (const auto &result : matches)
    {
        if (std::find(splittedText.begin(), splittedText.end(), result) != splittedText.end())
        {
            std::cout << result << std::endl;
        }
    }
}

void PowerSign(const std::string &pattern, const std::string &text)
{
    // 'text' kelimelerini ayırmak için boşlukla bölüyoruz
    std::istringstream textStream(text);
    std::vector<std::string> splittedText;
    std::string word;

    // Text'i kelimelere ayırıyoruz
    while (textStream >> word)
    {
        splittedText.push_back(word);
    }

    // 'pattern' içinde '^' karakterinden sonraki kısmı alıyoruz
    size_t caretPos = pattern.find('^');
    std::string startPattern;
    if (caretPos != std::string::npos)
    {
        startPattern = pattern.substr(caretPos + 1);
    }

    // Eşleşen kelimeleri tutmak için liste oluşturuyoruz
    std::vector<std::string> matches;

    // Tüm kelimeleri kontrol ediyoruz
    for (const auto &w : splittedText)
    {
        // Eğer kelime başlangıç paternine uyuyorsa, listeye ekliyoruz
        if (w.find(startPattern) == 0)
        {
            matches.push_back(w);
        }
    }

    // Eşleşen kelimeleri yazdırıyoruz
    for (const auto &match : matches)
    {
        std::cout << match << std::endl;
    }
}

void DollarSign(const std::string &pattern, const std::string &text)
{
    // 'text' içindeki kelimeleri boşlukla ayırın
    std::istringstream textStream(text);
    std::vector<std::string> splittedText;
    std::string word;

    // Text'i kelimelere ayırın
    while (textStream >> word)
    {
        splittedText.push_back(word);
    }

    // 'pattern' içinde '$' işaretinden önceki kısmı alın
    size_t dollarPos = pattern.find('$');
    std::string endPattern;
    if (dollarPos != std::string::npos)
    {
        endPattern = pattern.substr(0, dollarPos);
    }

    // Eşleşen kelimeleri tutmak için liste oluşturun
    std::vector<std::string> matches;

    // Her kelimeyi kontrol edin
    for (const auto &w : splittedText)
    {
        // Eğer kelime bitiş paternine uyuyorsa, listeye ekleyin
        if (w.size() >= endPattern.size() && w.compare(w.size() - endPattern.size(), endPattern.size(), endPattern) == 0)
        {
            matches.push_back(w);
        }
    }

    // Eşleşen kelimeleri yazdırın
    for (const auto &match : matches)
    {
        std::cout << match << std::endl;
    }
}

void PointSign(const std::string &pattern, const std::string &text)
{
    // 'text' içindeki kelimeleri boşluklarla ayırın
    std::istringstream textStream(text);
    std::vector<std::string> words;
    std::string word;

    // Text'i kelimelere ayırıyoruz
    while (textStream >> word)
    {
        words.push_back(word);
    }

    // 'pattern' içindeki '.' işaretine göre bölüyoruz
    size_t dotPos = pattern.find('.');
    std::string startPattern;
    std::string endPattern;

    if (dotPos != std::string::npos)
    {
        startPattern = pattern.substr(0, dotPos);
        endPattern = pattern.substr(dotPos + 1);
    }

    // Eşleşen kelimeleri tutmak için bir vektör oluşturuyoruz
    std::vector<std::string> matches;

    // Tüm kelimeleri kontrol ediyoruz
    for (const auto &w : words)
    {
        // Başlangıç ve bitiş desenlerine uyan ve uzunluğu aynı olan kelimeleri seçiyoruz
        if (w.size() == pattern.size() && w.find(startPattern) == 0 && w.rfind(endPattern) == w.size() - endPattern.size())
        {
            matches.push_back(w);
        }
    }

    // Eşleşen kelimeleri yazdırıyoruz
    for (const auto &match : matches)
    {
        std::cout << match << std::endl;
    }
}

void parser(const std::string &pattern, const std::string &text)
{
    std::vector<char> metacharacters = {'?', '|', '*', '+', '{', '[', '(', '^', '$', '.'};
    for (char c : pattern)
    {
        // Eğer karakter metakarakterler listesinde ise, uygun fonksiyonu çağırın
        if (std::find(metacharacters.begin(), metacharacters.end(), c) != metacharacters.end())
        {
            switch (c)
            {
            case '?':
                QuestionMark(pattern, text);
                break;
            case '|':
                OR(pattern, text);
                break;
            case '*':
                StarSign(pattern, text);
                break;
            case '+':
                PlusSign(pattern, text);
                break;
            case '{':
                CurlyBracket(pattern, text);
                break;
            case '[':
                SquareBracket(pattern, text);
                break;
            case '(':
                Bracket(pattern, text);
                break;
            case '^':
                PowerSign(pattern, text);
                break;
            case '$':
                DollarSign(pattern, text);
                break;
            case '.':
                PointSign(pattern, text);
                break;
            }
        }
    }
}

int main()
{
    std::string pattern, text;
    // Kullanıcıdan 'pattern' ve 'text' girdisini al
    std::cout << "Pattern giriniz: ";
    std::getline(std::cin, pattern); // Kullanıcıdan 'pattern' girdi alır
    std::cout << "Text giriniz: ";
    std::getline(std::cin, text); // Kullanıcıdan 'text' girdi alır

    // 'parser' fonksiyonunu çağır
    parser(pattern, text);

    return 0;
}